`credentials` Credentials Offer parameter values identifying `credentials_supported` object(s) in the issuer's metadata. (#86)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

This specification defines the following parameters for the Credential Offer object:

* `credential_issuer`: REQUIRED. The URL of the Credential Issuer, as defined in (#credential-issuer-identifier), from which the Wallet is requested to obtain one or more Credentials. The Wallet uses it to obtain the Credential Issuer's Metadata following the steps defined in (#credential-issuer-wellknown).

> * `credentials`: REQUIRED. An array of strings that each identify one of the keys in the key/value pairs stored in the `credentials_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.

Reviewer: ```suggestion
* `credentials`: REQUIRED. An array of unique strings that each identify one of the keys in the key/value pairs stored in the `credentials_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.
```

Author: when saying unique, I think we need to be a little more specific..? I think we would want `unique within that issuer, but not globally unique` ?

Reviewer 2: I suggest replacing `key` with `name` in when refering to members of JSON objects. I.e. `name/value pairs` instead of `key/value` pairs. Both https://www.json.org and https://www.rfc-editor.org/rfc/rfc8259.html seem to use the `name` terminology.

Reviewer: @Sakurann Yeah, but that is more or less assumed by just adding unique. But always better to be very explicit about it

Author: ```suggestion
* `credentials`: REQUIRED. An array of unique strings that each identify one of the keys in the name/value pairs stored in the `credentials_supported` Credential Issuer metadata property. The Wallet uses this string value to obtain the respective object that contains information about the Credential being offered as defined in (#credential-issuer-parameters). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.
```
agreed to change to `name/value pairs` and to keep `unique` (don't have a better text proposal to clarify)

> * `credentials`: REQUIRED. A JSON array of JSON strings that each uniquely identify one of the objects in the `credentials_supported` array included in the Credential Issuer metadata as defined in (#credential-issuer-parameters). The Wallet resolves this string value to the respective object to obtain information about the Credential being offered defined in (credential-metadata-object). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.

Reviewer 3: Could this be reworded to make it clear which element in the `credentials_supported` object it is supposed to match against? I also think it is redundant to say `uniquely identify` here, as the definition of the identifier already includes that.

Maybe something like:
```suggestion
* `credentials`: REQUIRED. A JSON array of JSON strings that each can be matched against the `credentials_supported_identifier` parameter of the objects in the `credentials_supported` array included in the Credential Issuer metadata as defined in (#credential-issuer-parameters). The Wallet resolves this string value to the respective object to identify and obtain information about the Credential being offered defined in (credential-metadata-object). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.
```

Author: ```suggestion
* `credentials`: REQUIRED. An array of strings that each can be matched against the `credentials_supported_identifier` parameter of the objects in the `credentials_supported` array included in the Credential Issuer metadata as defined in (#credential-issuer-parameters). The Wallet resolves this string value to the respective object to identify and obtain information about the Credential being offered defined in (credential-metadata-object). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.
```

Author: the structure has changed to map, please check an updated language..

> * `credentials`: REQUIRED. A JSON array of JSON strings that each uniquely identify one of the objects in the `credentials_supported` array included in the Credential Issuer metadata as defined in (#credential-issuer-parameters). The Wallet resolves this string value to the respective object to obtain information about the Credential being offered defined in (credential-metadata-object). For example, this string value can be used to obtain `scope` value to be used in the Authorization Request.

Author: fixed

In addition to a mechanism defined in (#credential-authz-request), Credential Issuers MAY support requesting authorization to issue a credential using OAuth 2.0 scope parameter.

> When the Wallet does not know which scope value to use to request issuance of a certain credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials` parameter values to identify object(s) in the `credentials_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.

Reviewer: Previously it allowed for the param to either being a string, which then needed to be used as a scope, or it was the object, from which you then had to lookup the scope in the respective `credentials_supported` object.

scopes are however completely optional in `credentials_supported`. Since the string will now always map onto the respective key of the `credentials_supported` map, this change means we now always need to look at the `scrope` value. Is that desired, or should the text be reworked that you need to lookup the scope and if not present use the string value?

Author: I think it is a desired/intended behavior that if the wallet needs to look up the scope value, it uses the identifier to look it up. if it does not need scope value, it just uses the parameters other than scope values to formulate authorization_details request. allowing to use identifier as a scope value will lead to ambiguities when that is allowed and when it is not..

Reviewer: Agreed, just pointing out the difference with previous version

> When the Wallet does not know which scope value to use to request issuance of a certain credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials` parameter values to identify object(s) in the `credentials_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.

Reviewer 2: Suggest "`credentials_supported` object in" instead of "`credentials_supported` map in" (i.e. use `object` instead of `map`).

Author: it used to say object and Timo suggested to change to map and I agreed. happy to change back if this comes up again..

> When the Wallet does not know which scope value to use to request issuance of a certain credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials` parameter values to identify object(s) in the `credentials_supported` array in the Credential Issuer metadata parameter and use `scope` parameter value from that object.

Reviewer 4: credentials_supported structure has been changed to an object/map

```suggestion
When the Wallet does not know which scope value to use to request issuance of a certain credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials` parameter values to identify object(s) in the `credentials_supported` object in the Credential Issuer metadata parameter and use `scope` parameter value from that object.
```

```suggestion
When the Wallet does not know which scope value to use to request issuance of a certain credential, it can discover it using the `scope` Credential Issuer metadata parameter defined in (#credential-issuer-parameters). When the flow starts with a Credential Offer, the Wallet can use the `credentials` parameter values to identify object(s) in the `credentials_supported` map in the Credential Issuer metadata parameter and use `scope` parameter value from that object.
```

Author: good catch - thank you!

* `require_credential_response_encryption`: OPTIONAL. If `require_credential_response_encryption` is `true`, the Credential Issuer requires additional encryption on top of TLS for the Credential Response and expects encryption parameters to be present in the Credential Request and/or Batch Credential Request. If `require_credential_response_encryption` is `false` or omitted, the Credential Issuer indicates that no additional Credential Response encryption is required. If `require_credential_response_encryption` is `true`, `credential_response_encryption_alg_values_supported` MUST also be provided.
* `display`: OPTIONAL. An array of objects, where each object contains display properties of a Credential Issuer for a certain language. Below is a non-exhaustive list of valid parameters that MAY be included:
  * `name`: OPTIONAL. String value of a display name for the Credential Issuer.
  * `locale`: OPTIONAL. String value that identifies the language of this object represented as a language tag taken from values defined in BCP47 [@!RFC5646]. There MUST be only one object for each language identifier.

> * `credentials_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is an identifier of the credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific credential and contains the following parameters defined by this specification:

Reviewer 5: If we have a breaking change here, it is a convenient opportunity for any renaming from #91 . This would also enable some backwards-compatiblity

> * `credentials_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is an identifier of the credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific credential and contains the following parameters defined by this specification:

Reviewer: ```suggestion
* `credentials_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is a unique identifier of the credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific credential and contains the following parameters defined by this specification:
```

> * `credentials_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is an identifier of the credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific credential and contains the following parameters defined by this specification:

Reviewer 2: `identifier of the credential being described` may collide with the different `credential_identifier` that we added in a different PR.
Perhaps `identifier of the supported credential being described`.

Author: ```suggestion
* `credentials_supported`: REQUIRED. An object that describes specifics of the Credential that the Credential Issuer supports issuance of. This object contains a list of name/value pairs, where each name is a unique identifier of the supported credential being described. This identifier is used in the Credential Offer as defined in (#credential_offer_parameters) to communicate to the Wallet which Credential is being offered. The value is an object that contains metadata about specific credential and contains the following parameters defined by this specification:
```

This section defines the structure of the objects that appear in the `credentials_supported` metadata parameter.

* `format`: REQUIRED. A JSON string identifying the format of this credential, i.e., `jwt_vc_json` or `ldp_vc`. Depending on the format value, the object contains further elements defining the type and (optionally) particular claims the credential MAY contain and information about how to display the credential. (#format_profiles) defines Credential Format Profiles introduced by this specification.

> * `credentials_supported_identifier`: REQUIRED when Credential Offer is used. A JSON string uniquely identifying this object within the `credentials_supported` array. It is used in the Credential Offer as defined in (#credential_offer_parameters) to enable the Wallet to discover the information about the Credential being offered.

Reviewer 3: Since this is a unique identifier could we consider renaming this to `id` and moving it above `format` in this definition to signify the importance?

Author: Within VCI, there are two different identifiers: 1) identifier that points to a credentials_supported object in the issuer metadata, and 2) identifier that identifies a credential instance during issuance (PR #65), so I was trying to differentiate those.

Having said that, there is only one identifier within issuer metadata, so maybe just "id" is ok...

There are so many "id" claims in the Credential space: in the DIF presentation exchange, in W3C VCDM, that i would prefer differentiating from those, but maybe no need...

On the order, I usually like putting mandatory parameters at the top, but in this case, logically, identifier should probably be at the top.

Reviewer 5: I agree with @cobward , it is the primary identifier and belongs to the top. I was also wondering why its not named `id` but I'm fine with the explanation

Author: with `credentials_subject` structure changing to a map, there will be no more `credentials_supported_identifier` :)
@cobward I think this data structure change of a `credentials_supported_identifier` is a breaking change for your implementation.. please check.

   [[ To be removed from the final specification ]]
   
   -19
  
   * changed the structure of the `credentials_supported` parameter to a map from array of objects

>    * changed the structure of `credentials` parameter in Credential Offer to only be a string (no more objects) whose value is a key in the `credentials_supported` list entry

Reviewer 4: ```suggestion
   * changed the structure of `credentials` parameter in Credential Offer to only be a string (no more objects) whose value is a key in the `credentials_supported` map
```
